### PR TITLE
methods are added to controller for crud and additional methods, dele…

### PR DESCRIPTION
**Pull request by krcg93**
## Summary
- methods are added to controller for crud and additional methods, delete method is added to model, user exceptions are added
## Trello Card
https://trello.com/c/u9l0nUuI/21-crear-user-controller-con-las-operaciones-crud-de-user

### CODE REVIEW

File: src/main/java/wolox/training/controllers/UserController.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * This method return one {@link User} by its username
+     *
+     * @param username: username
+     * @return {@link User} by username
+     */
+    @GetMapping("/{username}")
+    public User findByUsername(@PathVariable String username) {
+        return userRepository.findByUsername(username);
```
Comment (JaimLus) [inline src/main/java/wolox/training/controllers/UserController.java:39]: What happens if not find it?

File: src/main/java/wolox/training/controllers/UserController.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * This method delete an {@link Book} with following parameters
+     *
+     * @param book: book to delete
+     * @param id: id to search
+     * @return user with to deleted book
+     */
+    @DeleteMapping("/{id}/books")
+    public User deleteBook(@PathVariable Long id, @RequestBody Book book)
+            throws UserNotFoundException {
```
Comment (JaimLus) [inline src/main/java/wolox/training/controllers/UserController.java:120]: I think you don't need to have a throw exception.
The exception should extend from RuntimeException.

File: src/main/java/wolox/training/controllers/UserController.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+     * This method add an {@link Book} with following parameters
+     *
+     * @param book: book to add
+     * @param id: id to search
+     * @return user with to added book
+     */
+    @PutMapping("/{id}/books")
+    @ResponseStatus(HttpStatus.CREATED)
+    public User addBook(@PathVariable Long id, @RequestBody Book book)
+            throws UserNotFoundException {
```
Comment (JaimLus) [inline src/main/java/wolox/training/controllers/UserController.java:105]: Check the throw.

**Review by JaimLus — COMMENTED** [review]
Excellent job!

File: src/main/java/wolox/training/controllers/UserController.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    public User create(@RequestBody User user) {
+        return userRepository.save(user);
+    }
+
+    /**
+     * This method deletes an book by its id
+     *
+     * @param id: user id by delete
+     */
+    @DeleteMapping("/{id}")
```
Comment (juandbc) [inline src/main/java/wolox/training/controllers/UserController.java:71]: Here it should response with 204

File: src/main/java/wolox/training/controllers/UserController.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * This method replaces an {@link User} with following parameters
+     *
+     * @param user: user to save
+     * @param id: id to search
+     * @return user update
+     */
+    @PutMapping("/{id}")
+    public User updateUser(@RequestBody User user, @PathVariable Long id) {
```
Comment (juandbc) [inline src/main/java/wolox/training/controllers/UserController.java:86]: Here it should response with 204

File: src/main/java/wolox/training/controllers/UserController.java (hunk 3 above)
Comment (juandbc) [inline src/main/java/wolox/training/controllers/UserController.java:105]: It is not necessary the throws statement
```suggestion
    public User addBook(@PathVariable Long id, @RequestBody Book book) {
```

File: src/main/java/wolox/training/controllers/UserController.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+     * @param id: id to search
+     * @return user with to added book
+     */
+    @PutMapping("/{id}/books")
+    @ResponseStatus(HttpStatus.CREATED)
+    public User addBook(@PathVariable Long id, @RequestBody Book book)
+            throws UserNotFoundException {
+        User user = userRepository.findById(id).orElseThrow(UserNotFoundException::new);
+        user.addBook(book);
+        return userRepository.save(user);
```
Comment (juandbc) [inline src/main/java/wolox/training/controllers/UserController.java:108]: Good job throwing the exception when the user doesn't exist. But what happen when the book that you try to add is already added? It throws the exception but is missing the HTTP Status

**Review by juandbc — CHANGES_REQUESTED** [review]

**Pull request closed without merging**